File: src/sw/sa_pipeline/stage9_sa.py
class Stage9SA:
    """
    Ninth Pipe from SA_Verilog. This pipe is responsible to take the decision to do the swap or not.
    """

    def __init__(self):
        """

        """
        self.new_output: dict = {
            'th_idx': 0,
            'th_valid': False,
            'sw': False
        }
        self.old_output: dict = self.new_output.copy()

    def compute(self, st8_input: dict):
        """

        @param st8_input:
        """
        # moving forward the ready outputs
        self.old_output = self.new_output.copy()

        st8_th_idx = st8_input['th_idx']
        st8_th_valid = st8_input['th_valid']

        st8_dc = st8_input['dc']
        st8_ds = st8_input['ds']

        sw = st8_ds < st8_dc
        if sw:
            a=1

        self.new_output: dict = {
            'th_idx': st8_th_idx,
            'th_valid': st8_th_valid,
            'sw': sw
        }

File: src/cpp/sw/sa_pipeline/stage9_sa.cpp
#include "sa_pipeline_sw.h"

class Stage9SA {
private:
    bool sw;


public:
    ST9_OUT new_output = {0, false, false};
    ST9_OUT old_output = {0, false, false};

    Stage9SA() {
        this->sw = false;
    }

    void compute(ST8_OUT st8_input) {
        this->old_output.th_idx = this->new_output.th_idx;
        this->old_output.th_valid = this->new_output.th_valid;
        this->old_output.sw = this->new_output.sw;

        int st8_th_idx = st8_input.th_idx;
        bool st8_th_valid = st8_input.th_valid;
        int st8_dc = st8_input.dc;
        int st8_ds = st8_input.ds;

        bool sw_c = st8_ds < st8_dc;

        this->new_output.th_idx = st8_th_idx;
        this->new_output.th_valid = st8_th_valid;
        this->new_output.sw = sw_c;
    }
};